Add multivariate normal distribution

krstopro opened the pull request:
Added `multivariate_normal` to `Nx.Random`. Few things I am not sure about:
- Default method is `:cholesky`. Cholesky is also default for `JAX` while SVD is default for `NumPy`.
- The function does not check if the covariance matrix is positive semi-definite (PSD). I was considering implementing this as well, but it seemed a bit of an overkill. It is done in `NumPy`, but not in `JAX`. Also, Cholesky decomposition will fail if covariance is not PSD.

polvalente commented on nx/lib/nx/random.ex:
```diff
@@ -551,6 +551,181 @@ defmodule Nx.Random do
+
+        :eigh ->
+          {s, u} = Nx.LinAlg.eigh(covariance)
+          u * Nx.sqrt(Nx.abs(s))
```
Why take the sqrt of the eigenvalues here?

polvalente replied:
Makes sense!
You can add a line to the docs that specifies that if covariance isn't positive semi-definite, it will be garbage in, garbage out, and get rid of the abs

krstopro replied:
Done!

polvalente commented on nx/lib/nx/random.ex:
```diff
@@ -551,6 +551,181 @@ defmodule Nx.Random do
+      iex> {multivariate_normal, _new_key} = Nx.Random.multivariate_normal(key, Nx.tensor([0, 0]), Nx.tensor([[1, 0], [0, 1]]), shape: {3, 2})
+      iex> multivariate_normal
+      #Nx.Tensor<
+        f32[3][2][2]
```
I'm a bit confused about these output shapes. I guess they are related to the input tensor shapes, but given the rest of the behavior in Nx.Random, I'd expect the :shape option to fully control the output shape

krstopro replied:
The last dimension is equal to the dimension of the distribution (in this case 2, as both mean and covariance have dimensions 2). If shape option is not provided (i.e. `nil`), then the output will be of shape `{dim}`.
If shape option is `{3, 2}`, then the output will be of shape `{3, 2, 2}`.
This is the same as in NumPy and JAX.
https://github.com/numpy/numpy/blob/104addf377577d12ea34547303822369fcb6e44e/numpy/random/_generator.pyx#L3777
https://jax.readthedocs.io/en/latest/_autosummary/jax.random.multivariate_normal.html#jax.random.multivariate_normal

polvalente replied:
The docs can use some improvement to clarify this. Something along the lines of the first 3 phrases of this answer should suffice :)

krstopro replied:
Agreed!

krstopro replied:
And done!

krstopro commented:
@polvalente Alright, updated the function. Sorry it took me a while. I also updated one of the unit tests (the one for properties). Previously, off-diagonal values of the covariance were not estimated and tested. I guess this should be computed using `Nx.covariance` which we don't have at the moment.

polvalente commented on nx/lib/nx/random.ex:
```diff
@@ -551,6 +551,186 @@ defmodule Nx.Random do
+
+    * `:names` - the names of the returned tensor
+
+    * `:method` - a decomposition methond used for the covariance. Must be one of :svd, :eigh, and :cholesky.
```
```suggestion
    * `:method` - a decomposition method used for the covariance. Must be one of :svd, :eigh, and :cholesky.
```

polvalente commented on nx/lib/nx/random.ex:
```diff
@@ -551,6 +551,186 @@ defmodule Nx.Random do
+
+    * `:type` - a float type for the returned tensor
+
+    * `:shape` - batch shape of the returned tensor, i.e. the prefix of the result shape exluding the last axis
```
```suggestion
    * `:shape` - batch shape of the returned tensor, i.e. the prefix of the result shape excluding the last axis
```

polvalente commented on nx/lib/nx/random.ex:
```diff
@@ -551,6 +551,186 @@ defmodule Nx.Random do
+    * `:names` - the names of the returned tensor
+
+    * `:method` - a decomposition methond used for the covariance. Must be one of :svd, :eigh, and :cholesky.
+      Default :cholesky. For singular covariance matrices, use :svd or :eigh.
```
```suggestion
      Defaults to :cholesky. For singular covariance matrices, use :svd or :eigh.
```

polvalente commented:
> @polvalente Alright, updated the function. Sorry it took me a while. I also updated one of the unit tests (the one for properties). Previously, off-diagonal values of the covariance were not estimated and tested. I guess this should be computed using `Nx.covariance` which we don't have at the moment.

@krstopro We can add `covariance`. Let's move ahead with the PR as is. If you want to send a follow-up with covariance and replacing where appropriate here, it would be welcome!

krstopro commented:
> > @polvalente Alright, updated the function. Sorry it took me a while. I also updated one of the unit tests (the one for properties). Previously, off-diagonal values of the covariance were not estimated and tested. I guess this should be computed using `Nx.covariance` which we don't have at the moment.
> 
> @krstopro We can add `covariance`. Let's move ahead with the PR as is. If you want to send a follow-up with covariance and replacing where appropriate here, it would be welcome!

Sure, shouldn't be hard!

polvalente commented:
> Sure, shouldn't be hard!

Bold statement! :laughing:
